src/hotspot/cpu/aarch64/register_aarch64.hpp

```diff
@@ -34,42 +34,44 @@ typedef VMRegImpl* VMReg;
 
 // Use Register as shortcut
 class RegisterImpl;
-typedef RegisterImpl* Register;
+typedef const RegisterImpl* Register;
 
-inline const Register as_Register(int encoding) {
-  return (Register)(intptr_t) encoding;
-}
+inline constexpr Register as_Register(int encoding);
 
 class RegisterImpl: public AbstractRegisterImpl {
- public:
+  static constexpr Register first();
+
+public:
   enum {
     number_of_registers         =   32,
     number_of_byte_registers      = 32,
-    number_of_registers_for_jvmci = 34,  // Including SP and ZR.
+    number_of_declared_registers  = 34,  // Including SP and ZR.
     max_slots_per_register = 2
   };
 
   // derived registers, offsets, and addresses
-  Register successor() const                          { return as_Register(encoding() + 1); }
+  const Register successor() const { return this + 1; }
 
   // construction
-  inline friend const Register as_Register(int encoding);
+  inline friend constexpr Register as_Register(int encoding);
```

Comment:
This is getting bikesheddy so feel free to ignore:

Instead if using friend I would probably either just make `first()` public. 
Or, make as_xxx a static class method and wrap it with a global scope wrapper like this:

```
--- a/src/hotspot/cpu/aarch64/register_aarch64.hpp
+++ b/src/hotspot/cpu/aarch64/register_aarch64.hpp
@@ -53,7 +53,8 @@ public:
   const Register successor() const { return this + 1; }

   // construction
-  inline friend constexpr Register as_Register(int encoding);
+  static constexpr Register as(int encoding);
+
```
```
diff --git a/src/hotspot/share/asm/register.hpp b/src/hotspot/share/asm/register.hpp
index 06a8735f520..7d9036b0ff6 100644
--- a/src/hotspot/share/asm/register.hpp
+++ b/src/hotspot/share/asm/register.hpp
@@ -59,9 +59,12 @@ enum { name##_##type##EnumValue = (value) }
 #else // USE_POINTERS_TO_REGISTER_IMPL_ARRAY
 
 #define REGISTER_IMPL_DECLARATION(type, impl_type, reg_count)           \
-inline constexpr type as_ ## type(int encoding) {                       \
+inline constexpr type impl_type::as(int encoding) {                     \
   return impl_type::first() + encoding;                                 \
 }                                                                       \
+inline constexpr type as_ ## type(int encoding) {                       \
+  return impl_type::as(encoding);                                       \
+}                                                                       \
 extern impl_type all_ ## type ## s[reg_count + 1] INTERNAL_VISIBILITY;  \
 inline constexpr type impl_type::first() { return all_ ## type ## s + 1; }
```